multinet/metadata/utils.py

```diff
@@ -0,0 +1,134 @@
+"""Utilities for dealing with metadata."""
+
+import json
+from dateutil import parser as dateutilparser
+from datetime import datetime
+
+from multinet.validation import (
+    ValidationFailure,
+    IncompatibleMetadata,
+    MetadataColumnKeyNotFound,
+)
+from multinet.types import TableMetadata
+
+# Import types
+from typing import Any, Callable, List, Dict, Tuple, Optional, Union
+
+EntryProcessingFunction = Callable[[str], Any]
+
+
+def process_boolean_entry(entry: str) -> bool:
+    """Try to determine base format of boolean so it can be converted properly."""
+
+    def from_int(x: str) -> Optional[bool]:
+        if x == "0" or x == "1":
+            return bool(int(x))
+        return None
+
+    def from_json_bool(x: str) -> Optional[bool]:
+        if x == "true" or x == "false":
+            return json.loads(x)
+        return None
+
+    def from_yaml_bool(x: str) -> Optional[bool]:
+        if x == "no" or x == "off":
+            return False
+        if x == "yes" or x == "on":
+            return True
+
+        return None
+
+    def cast_col_entry(x: str) -> bool:
+        casters = [from_int, from_json_bool, from_yaml_bool]
+        for caster in casters:
+            cast_row = caster(x)
+            if cast_row is not None:
+                return cast_row
+
+        raise ValueError
+
+    return cast_col_entry(entry)
+
+
+def process_date_entry(entry: str) -> str:
+    """Try to read a date as an ISO 8601 string or unix timestamp."""
+
+    if entry.isdigit():
+        # Raises ValueError if the number is out of range
+        return datetime.fromtimestamp(int(entry)).isoformat()
```

Comment:
Good point. It also seems that Arango chooses to store timestamps with millisecond precision, whereas I think conventionally timestamps are specified in second precision, so we'll have to do a conversion. Unfortunately, I don't think there's a surefire way to know if a specified timestamp uses second or millisecond precision, but I think going with the convention is fine for now.

Reply:
My bad, the second part I mentioned is actually irrelevant, as we're not storing dates as timestamps in arango, but as datestrings. So I think we should still assume that incoming timestamps use second precision. Just wanted to clear up confusion.

Reply:
8969159

Reply:
If we store them as ISO date strings, then the precision can go up to microseconds, AFAICT:

```python
import datetime
datetime.datetime.now().isoformat()
```

gives me a result of `2020-11-25T10:00:03.144373`. Similarly, `datetime.datetime.fromtimestamp(0.001).isoformat()` gives me `1969-12-31T19:00:00.001000` and `datetime.datetime.fromtimestamp(0).isoformat()` gives `1969-12-31T19:00:00`. All of these strings parse nicely into JavaScript `Date` objects, so I think in general converting to float and pushing these through string conversion into Arango will work out just fine.

Is that the current situation?

Reply:
Yes that is the current situation, i was just stating that I think the units of the incoming timestamps should be seconds, since that's the most common way to specify timestamps. They can still be floats, so there's no loss of precision.

Reply:
Oh I think I understand. When you said "millisecond precision" you meant that, e.g., JavaScript timestamps are expressed in milliseconds (your use of "precision" had confused me). Standard timestamps do indeed come in *units* of seconds, but their *precision* is in microseconds due to the use of floats.

Ok I think we're on the same page and this was just a vocabulary issue. Thanks!